plugins/diff/erofs/differ.go

```diff
@@ -0,0 +1,192 @@
+*/
+
+package erofs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/v2/core/content"
+	"github.com/containerd/containerd/v2/core/diff"
+	"github.com/containerd/containerd/v2/core/images"
+	"github.com/containerd/containerd/v2/core/mount"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+var emptyDesc = ocispec.Descriptor{}
+
+// erofsDiff does erofs comparison and application
+type erofsDiff struct {
+	store         content.Store
+	mkfsExtraOpts []string
+}
+
+func NewErofsDiff(store content.Store) diff.Comparer {
+	return &erofsDiff{
+		store: store,
+	}
+}
+
+// Compare creates a diff between the given mounts and uploads the result
+// to the content store.
+func (s erofsDiff) Compare(ctx context.Context, lower, upper []mount.Mount, opts ...diff.Opt) (d ocispec.Descriptor, err error) {
+	return emptyDesc, fmt.Errorf("erofsDiff does not implement Compare method: %w", errdefs.ErrNotImplemented)
+}
+
+func NewErofsApplier(store content.Store, mkfsExtraOpts []string) diff.Applier {
+	return &erofsDiff{
+		store:         store,
+		mkfsExtraOpts: mkfsExtraOpts,
+	}
+}
+
+func (s erofsDiff) ConvertErofs(ctx context.Context, r io.Reader, layerPath string) error {
+	args := append([]string{"--tar=f", "--aufs", "--quiet", "-Enoinline_data"}, s.mkfsExtraOpts...)
+	args = append(args, layerPath)
+	cmd := exec.CommandContext(ctx, "mkfs.erofs", args...)
+	cmd.Stdin = r
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("erofs apply failed: %s: %w", out, err)
+	}
+	log.G(ctx).Infof("running %s %s %v", cmd.Path, cmd.Args, string(out))
+	return nil
+}
+
+// Get the snapshot layer directory in order to generate EROFS-formatted blobs;
+//
+// If mount[0].Type is `bind` or `erofs`, it just tries the source dir; Or if
+// mount[0].Type is `overlayfs`, it tries the parent of the upperdir;
+//
+// The candidate will be checked with ".erofslayer" to make sure this active
+// snapshot is really generated by the EROFS snapshotter instead of others.
+func mountsToLayer(mounts []mount.Mount) (string, error) {
+	var layer string
+	mnt := mounts[0]
+	if mnt.Type == "bind" || mnt.Type == "erofs" {
+		layer = filepath.Dir(mnt.Source)
+	} else if mnt.Type == "overlay" {
+		layer = ""
+		for _, o := range mnt.Options {
+			if strings.HasPrefix(o, "upperdir=") {
+				layer = filepath.Dir(strings.TrimPrefix(o, "upperdir="))
+			}
+		}
+		if layer == "" {
+			return "", fmt.Errorf("unsupported overlay layer for erofs differ: %w", errdefs.ErrNotImplemented)
+		}
+	} else {
+		return "", fmt.Errorf("invalid filesystem type for erofs differ: %w", errdefs.ErrNotImplemented)
+	}
+	// If the layer is not prepared by the EROFS snapshotter, fall back to the next differ
+	if _, err := os.Stat(filepath.Join(layer, ".erofslayer")); err != nil {
```

Comment:
This looks difficult to maintain... I think applier shouldn't be aware of internal directory structure of a snapshotter.

Reply:
> This looks difficult to maintain... I think applier shouldn't be aware of internal directory structure of a snapshotter.

>  I think applier shouldn't be aware of internal directory structure of a snapshotter.
For the generic differ I think, yes.  But this differ is only useful together with erofs snapshotter.
All cases are as below:
erofs snapshot layer + erofs differ : convert media type
erofs snapshot layer + other differs (walking for example) : just unpack to `fs` and behaves as overlay;
other snapshot layer (e.g. overlay or remote snapshotter) + erofs differ: move to the next differ;
other snapshot layer + other differ : don't care.

> This looks difficult to maintain...
Why? Basically it needs a way to identify a layer prepared by erofs snapshotter to ensure the safety.
Do you have better ideas for this requirement?..
I've checked the previous windows differs, but they seems they just use a special mount structure. But I think this method is more clean compared to that since the mount structure can be mounted directly (so that erofs snapshotter can be used with walking differ) rather than a difficult-to-identify mount structure...

Reply:
...

> 
> > This looks difficult to maintain...
> 
> Why? Basically it needs a way to identify a layer prepared by erofs snapshotter to ensure the safety. Do you have better ideas for this requirement?.. I've checked the previous windows differs, but they seems they just use a special mount structure. But I think this method is more clean compared to that since the mount structure can be mounted directly (so that erofs snapshotter can be used with walking differ) rather than a difficult-to-identify mount structure...

Derek has an [ongoing mount manager](https://github.com/dmcgowan/containerd/pull/6) which could help us to clean up this with new mount[] structure too, I think after this work lands, EROFS snapshotter could have a more clean way then.
But I guess they could be done in parallel? I do hope if you could agree..